shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm

```diff
@@ -1286,7 +1286,10 @@ - (NSString*)textInRange:(UITextRange*)range {
   NSAssert([range isKindOfClass:[FlutterTextRange class]],
            @"Expected a FlutterTextRange for range (got %@).", [range class]);
   NSRange textRange = ((FlutterTextRange*)range).range;
-  NSAssert(textRange.location != NSNotFound, @"Expected a valid text range.");
+  if (textRange.location == NSNotFound) {
+    /// Avoids [crashes](https://github.com/flutter/flutter/issues/138464) from an assertion against NSNotFound.
```

Comment:
the range is created by our code since it is typed FlutterTextRange. So I suspect something more is going on. Do you have a repro? If not, we should at least do a NSLog to produce a warning

Reply:
@importRyan Can we trace back to where this textRange of `NSNotFound` is created?

Reply:
[Here](https://github.com/flutter/engine/pull/16496/files) I think? Returning `NSNotFound` seems legit.

Reply:
[Here](https://github.com/flutter/engine/pull/16496/files) I think? Returning NSNotFound seems legit.> 

The helper function makes sense to return `NSNotFound`, and it's a good lead - there are a 5-6 usage of this `RangeForCharacterAtIndex` function in our text input plugin, and we should be able to trace to which one (e.g. set an associated object), and figure out why we are getting this `NSNotFound`

Reply:
It could be legit or an unaccounted for platform behavior that needs alternative handling to comply with the user intent, such that nil might be a suboptimal return.

As an engineer, it interests me to solve the root issue. With my product hat on, though, stopping a needless crash for our users is the thing I'd hoped to get done in as few days as possible. This issue's earlier reports predate this PR by a while.

Here is a sample trace that's identical for many users on iOS 17 this week in Crashlytics, although without some additional context on what `setTextInputState` did it's not terribly helpful. I (may errantly) assume there is some interplay between what's called by the system for actions around autofill, the configuration for the field, and the ranging finding. 

```
Fatal Exception: NSInternalInconsistencyException
0  CoreFoundation                 0x83f20 __exceptionPreprocess
1  libobjc.A.dylib                0x172b8 objc_exception_throw
2  Foundation                     0x6de7f8 _userInfoForFileAndLine
3  Flutter                        0x43c34 -[FlutterTextInputView textInRange:] + 1287 (FlutterTextInputPlugin.mm:1287)
4  UIKitCore                      0x4f9188 -[UIKBRTIPartner _queryUIKitDocumentRequest:completion:]
5  UIKitCore                      0x500aa8 -[UIKBRTIPartner _queryDocumentRequest:completion:]
6  UIKitCore                      0x5009a4 -[UIKBRTIPartner _updateRTIStateWithCompletion:]
7  UIKitCore                      0x5001f4 -[UIKBRTIPartner updateStateWithCompletion:updateTraits:]
8  UIKitCore                      0x3a0488 -[UIKBRTIPartner documentStateChanged:]
9  UIKitCore                      0xcf81e0 __37-[UIKeyboardImpl selectionDidChange:]_block_invoke_3
10 UIKitCore                      0x4e1710 -[UIKeyboardTaskEntry execute:]
11 UIKitCore                      0x4e15a4 -[UIKeyboardTaskQueue continueExecutionOnMainThread]
12 UIKitCore                      0x4e1210 -[UIKeyboardTaskQueue addTask:]
13 UIKitCore                      0xcf7ea0 -[UIKeyboardImpl selectionDidChange:]
14 Flutter                        0x4270c -[FlutterTextInputView setTextInputState:] + 1053 (FlutterTextInputPlugin.mm:1053)
15 Flutter                        0x4ad84 -[FlutterTextInputPlugin handleMethodCall:result:] + 2405 (FlutterTextInputPlugin.mm:2405)
16 Flutter                        0x5d413c __45-[FlutterMethodChannel setMethodCallHandler:]_block_invoke + 322 (FlutterChannels.mm:322)
17 Flutter                        0x5eaec invocation function for block in flutter::PlatformMessageHandlerIos::HandlePlatformMessage(std::_fl::unique_ptr<flutter::PlatformMessage, std::_fl::default_delete<flutter::PlatformMessage>>) + 110 (ref_ptr.h:110)
18 libdispatch.dylib              0x213c _dispatch_call_block_and_release
19 libdispatch.dylib              0x3dd4 _dispatch_client_callout
20 libdispatch.dylib              0x125a4 _dispatch_main_queue_drain
21 libdispatch.dylib              0x121b8 _dispatch_main_queue_callback_4CF
22 CoreFoundation                 0x56710 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__
23 CoreFoundation                 0x53914 __CFRunLoopRun
24 CoreFoundation                 0x52cd8 CFRunLoopRunSpecific
25 GraphicsServices               0x11a8 GSEventRunModal
26 UIKitCore                      0x40aae8 -[UIApplication _run]
27 UIKitCore                      0x4bed98 UIApplicationMain
28 UIKitCore                      0x638504 keypath_get_selector_hoverStyle
29 Runner                         0x868c main (AppDelegate.swift)
30 ???                            0x1ad8eb154 (Missing)
```

Reply:
so looks like it may be a result of framework textfield update that may provide a selection range outside of the textinputplugin. it seems to match the description on this pr

https://github.com/flutter/engine/pull/16496

It is probably when cursor is at the end of this. I am not sure why it doesn't crash after the change though. It seems it should be reliably reproduce if the cursor is at the end.

Reply:
> I am not sure why it doesn't crash after the change though

Probably because Apple's implementation is defensive enough to handle nil return for the `textForRange` API, which i'm not surprised. 

It sounds like we are indeed feeding the wrong range (from the framework). @importRyan Are you able to identify which usage of `RangeForCharacterAtIndex` in the plugin created such range? (from my previous comment?)

Reply:
After inspecting the Swift signatures of the [UITextInput](https://developer.apple.com/documentation/uikit/uitextinput#1652995) protocol, it seems to me the protocol demands that a non-nil `UITextRange` or `UITextPosition` value must be valid in the text, and the `UITextInput` implementation shouldn't construct an invalid text range (but it's funny that the objective-c docs don't say much about when / whether you should return nil for these methods, except in `positionFromPosition:offset:` and `positionFromPosition:inDirection:offset:`).

This won't entirely prevent the keyboard implementation from calling the `UITextInput` implementation with completely bogus ranges and positions (for instance they keyboard may call `selectedTextRange`, cache the value, and proceed to use the now invalid range after the text changes), but as @hellohuanlin pointed out, a `FlutterTextRange` with a `NSRange` containing `NSNotFound` seems to violate the protocol.

So I feel it would be better to vet the implementation of `func textRange(from fromPosition: to toPosition: ) -> UITextRange?` et.al and make sure they return nil instead of an invalid range / position (so the keyboard implementation gets to know if the range is actually invalid). That said I'm ok with just properly handling `NSNotFound` for now, file an issue to revise those UITextInput methods later.

Reply:
> That said I'm ok with just properly handling NSNotFound for now, file an issue to revise those UITextInput methods later.

SGTM. I'd also request adding some comment explaining this workaround and also link to the issues to be filed. @importRyan are you able to make the changes?

Reply:
That rather jives with my read of the delegate definition. While a UITextRange can be empty, location `NSNotFound: Int` is also a signal checked for by subclassed UITextField implementations, as well as when using some [system text wrangling APIs](https://developer.apple.com/library/archive/documentation/StringsTextFonts/Conceptual/TextAndWebiPhoneOS/LowerLevelText-HandlingTechnologies/LowerLevelText-HandlingTechnologies.html).

I'm a little lost, though, what are the specific actions you're seeking? 
Do they include:
a. Opening a new issue to patch the underlying range finding?
b. Commenting where NSAsserts were deleted and mentioning #138464? (done already)
c. Proper handling of NSNotFound in this PR's modified methods or elsewhere?

I'm not sure what the latter might mean in this context. 

If it's changing details of range finding methods (e.g., modifying `textRange(from, to:)`), I am regrettably without sufficient time this month to do a proper job. To prevent regressions, I'd need likely need to stand up several new tests. (I may have missed a battery against a full logical range of selection possibilities, starting states, perturbances, and character widths.) I'd need to take some sit down time to take in the full system and idiosyncrasies of some string indexing objects I haven't used (e.g., the tokenizer warned about in `textRange(from, to:)`).

As I'd love to stop those introduced NSAsserts from crashing MacroFactor due to valid user input, perhaps there's an interim ground by introducing an NSLog here? That could lower the bar a bit for other developers to participate in fixing the system.

Reply:
@importRyan I filed the issue https://github.com/flutter/flutter/issues/160100. Could you add a TODO and link to this issue in the code comment? Then we are good to go!

Reply:
Done! Thanks for teeing that up.